toqito/state_props/is_separable.py

```diff
@@ -66,247 +207,550 @@ def is_separable(state: np.ndarray, dim: None | int | list[int] = None, level: i
-                        p[0, 1, 7, 8] + p[1, 2, 6, 7],
-                        p[0, 2, 7, 8] + p[1, 2, 6, 8],
-                    ],
-                    [
-                        p[3, 4, 6, 7],
-                        p[3, 5, 6, 8],
-                        p[4, 5, 7, 8],
-                        p[3, 4, 6, 8] + p[3, 5, 6, 7],
-                        p[3, 4, 7, 8] + p[4, 5, 6, 7],
-                        p[3, 5, 7, 8] + p[4, 5, 6, 8],
-                    ],
-                    [
-                        p[0, 1, 3, 7] - p[0, 1, 4, 6],
-                        p[0, 2, 3, 8] - p[0, 2, 5, 6],
-                        p[1, 2, 4, 8] - p[1, 2, 5, 7],
-                        p[0, 1, 3, 8] - p[0, 1, 5, 6] + p[0, 2, 3, 7] - p[0, 2, 4, 6],
-                        p[0, 1, 4, 8] - p[0, 1, 5, 7] + p[1, 2, 3, 7] - p[1, 2, 4, 6],
-                        p[0, 2, 4, 8] - p[0, 2, 5, 7] + p[1, 2, 3, 8] - p[1, 2, 5, 6],
-                    ],
-                    [
-                        p[0, 3, 4, 7] - p[1, 3, 4, 6],
-                        p[0, 3, 5, 8] - p[2, 3, 5, 6],
-                        p[1, 4, 5, 8] - p[2, 4, 5, 7],
-                        p[0, 3, 4, 8] - p[1, 3, 5, 6] + p[0, 3, 5, 7] - p[2, 3, 4, 6],
-                        p[0, 4, 5, 7] - p[1, 4, 5, 6] + p[1, 3, 4, 8] - p[2, 3, 4, 7],
-                        p[0, 4, 5, 8] - p[2, 3, 5, 7] + p[1, 3, 5, 8] - p[2, 4, 5, 6],
-                    ],
-                    [
-                        p[0, 4, 6, 7] - p[1, 3, 6, 7],
-                        p[0, 5, 6, 8] - p[2, 3, 6, 8],
-                        p[1, 5, 7, 8] - p[2, 4, 7, 8],
-                        p[0, 4, 6, 8] - p[1, 3, 6, 8] + p[0, 5, 6, 7] - p[2, 3, 6, 7],
-                        p[0, 4, 7, 8] - p[1, 3, 7, 8] + p[1, 5, 6, 7] - p[2, 4, 6, 7],
-                        p[0, 5, 7, 8] - p[2, 3, 7, 8] + p[1, 5, 6, 8] - p[2, 4, 6, 8],
-                    ],
-                ]
+    """
+    # --- 1. Input Validation, Normalization, Dimension Setup ---
+    if not isinstance(state, np.ndarray):
+        raise TypeError("Input state must be a NumPy array.")
+    if state.ndim != 2 or state.shape[0] != state.shape[1]:
+        raise ValueError("Input state must be a square matrix.")
+
+    if np.issubdtype(state.dtype, np.complexfloating):
+        machine_eps = np.finfo(state.real.dtype).eps
+    elif np.issubdtype(state.dtype, np.floating):
+        machine_eps = np.finfo(state.dtype).eps
+    else:
+        machine_eps = np.finfo(float).eps
+
+    state_len = state.shape[0]
+
+    if not is_positive_semidefinite(state, atol=tol, rtol=tol):
+        raise ValueError("Checking separability of non-positive semidefinite matrix is invalid.")
+
+    trace_state_val = np.trace(state)
+    current_state = state.copy()
+
+    if state_len > 0 and abs(trace_state_val) < tol:
+        if np.any(
+            np.abs(current_state)
+            > 100 * machine_eps * max(1, np.max(np.abs(current_state)) if current_state.size > 0 else 1)
```

Comment:
For the constant of `100`, should we note anywhere why we choose this as a "magic" constant number? I'm fine with it, but I wonder if we can either define it in a variable and/or add a comment as to what this number is and why it is defined as such.